feat(#4): add linter to ci server

Celleforst opened the pull request:
This PR aims to add the `run_linter_check` function and unittests in order to close issue #4.

Strengthless committed:
feat(#8): add file operation functions in utils.py

Strengthless committed:
feat(#13): implement skeletons for fastAPI router and all modules

Strengthless committed:
chore(#13): clean up unused packages, e.g. fastgithub

Celleforst committed:
Merge branch '13-implement-skeleton-for-the-fastapi-router' into 4-add-linter-to-ci-server

Strengthless committed:
Merge branch 'main' into 13-implement-skeleton-for-the-fastapi-router

Celleforst committed:
feat(#4): Implement linter function

Implement linter function requested by issue #4. Linter function will be
used to check for syntax errors in all files in the repo.

Celleforst committed:
test(#4): Add unittests for linting function

Adds two unittests to run_linter_check function (one successful and one
failure case)

Strengthless committed:
chore(#13): migrate to use Uvicorn instead of FastAPI CLI

Strengthless committed:
Merge remote-tracking branch 'origin/12-implement-skeleton-for-the-fastapi-router' into 4-add-linter-to-ci-server

Strengthless commented on tests/test_actions.py:
```diff
@@ -0,0 +1,22 @@
+class ActionsTest(unittest.TestCase):
+
+    # Tests for run_linter_check
+    def test_linting_without_errors(self):
+        ephermal_folder = "./temp/actions_test/"
+        write_to_file(ephermal_folder + "hello.py", "print('hello world!')\n\n")
+        output = run_linter_check(ephermal_folder)
+        self.assertTrue(output[0])
+
+    def test_linting_with_errors(self):
+        ephermal_folder = "./temp/actions_test/"
+        write_to_file(ephermal_folder + "hello.py", "print('hello world!)")
+        output = run_linter_check(ephermal_folder)
+        self.assertFalse(output[0])
+        self.assertTrue("E999 SyntaxError" in output[1])
```
Instead of creating ephemeral folders and writing files each time. 

How about we just create static test fixtures and invoke them? Something like `tests/fixtures/syntax_correct.py` and `tests/fixtures/syntax_error.py` could work.

Strengthless commented on tests/test_actions.py:
```diff
@@ -0,0 +1,22 @@
+
+class ActionsTest(unittest.TestCase):
+
+    # Tests for run_linter_check
```
In addition to those two tests, we can also add one more that makes sure the linter does not report styling "errors", e.g. unused imports or variables, or indentation contains mixed spaces and tabs.

Strengthless commented on src/modules/actions.py:
```diff
@@ -44,11 +48,22 @@ def run_linter_check(target_folder: str) -> tuple[bool, str]:
+    if not check_if_folder_exists(target_folder):
+        raise ValueError(f"The provided path {target_folder} is not a valid directory.")
+
+    flake_command = "flake8 --select E999 ."
```
I just took a deeper look into this, and perhaps we need to include a few more error codes:

We can probably reference [this article](https://michaelcurrin.github.io/dev-cheatsheets/cheatsheets/python/linting/flake8.html), as it looks good.

E9 -> This includes E999 from `flake8`, and E901 E902 from `pycodestyle` (as mentioned in [flake8 documentation](https://flake8.pycqa.org/en/latest/user/error-codes.html), we also need to reference [pycodestyle’s list of error codes](https://pycodestyle.readthedocs.io/en/latest/intro.html#error-codes)). All of them should be reported, as they're obviously syntax errors that will cause problems.
<img width="696" alt="image" src="https://github.com/user-attachments/assets/a546f38b-bcf1-4f36-8f51-17a7ffbe7f0f" />
<img width="396" alt="image" src="https://github.com/user-attachments/assets/14995b9b-db2f-4963-91ec-257f776ea6b3" />

F63 -> Technically this is not a syntax error, as the python code probably still runs - but it's very likely that the written code does not behave as intended, and I think we should report this. It's totally up to you whether you include this though.
<img width="457" alt="image" src="https://github.com/user-attachments/assets/ace74bd7-662c-4434-bfd1-17c9658b4582" />

F7 -> syntax error related to iterations that will cause python to not run.
<img width="556" alt="image" src="https://github.com/user-attachments/assets/9ea81305-c27f-4042-8630-b5064e6efbcf" />

F82 -> runtime errors that will cause python to not run, should be reported
<img width="427" alt="image" src="https://github.com/user-attachments/assets/1a7c5f54-9163-4bd4-b49a-7132963b894b" />

Celleforst replied:
Sure I am adding all these checks. However, I tried making flake8 report E901 but it only reports E999... And I did not manage to trigger F63.

Celleforst replied:
> Sure I am adding all these checks. However, I tried making flake8 report E901 but it only reports E999... And I did not manage to trigger F63.

Apart from that, it should be ready to merge

Strengthless reviewed (changes requested):
Overall looks good! Just a few things that we'll need work on. Thanks @Celleforst

Celleforst committed:
Merge remote-tracking branch 'origin/main' into 4-add-linter-to-ci-server

Strengthless committed:
Merge branch 'main' into 4-add-linter-to-ci-server

Celleforst committed:
fix(#4): pre-commit checks ignore tests/fixtures folder

Make flake8 and black pre-commit actions ignore test/fixtures folder
since it may contain intentional errors

Celleforst committed:
Merge branch '4-add-linter-to-ci-server' of github.com:dd2480-spring-2025-group-1/assignment-2 into 4-add-linter-to-ci-server

Celleforst committed:
ref(#4): add fixtures folder for tests

Adds folder tests/fixtures to store static test fixtures to be used in
tests. Note that I did not manage to commit whilst ignoring errors in
the fixtures folder yet. Thus syntax_error.py does not contain any
errors for now.

Strengthless committed:
fix(#4): linter yelling because of the `tests/fixtures/syntax_error` folder

Celleforst committed:
fix(#4): add more linting checks

Adds more linting checks that are relevant to having a correct syntax

Celleforst committed:
fix(#4): Add more tests to linting and refactor structure

Adds more tests for linting function and add flake8_tests folder to
better organize fixture folder

Strengthless commented on .github/workflows/black.yml:
```diff
@@ -18,4 +18,4 @@ jobs:
       - name: Run format check
         run: |
-          python -m black --check .
+          python -m black --check --extend-exclude tests/fixtures .
```
Do we actually need this? 

Github workflows emulate the setup process and should be able to pick up the Black config file. This means that if `python -m black --check .` works on your machine, it's likely that it will also work the same for github workflows, and you probably don't need to include that flag. If anything, it should be in `pyproject.toml`

Celleforst replied:
@Strengthless when trying to commit the second sample for syntax_correct with no empty line at the end of the file, it would not let me commit without adding the line. Does that mean that it does not use the pyproject.toml?

Strengthless commented on .github/workflows/flake8.yml:
```diff
@@ -18,4 +18,4 @@ jobs:
       - name: Run lint check
         run: |
-          python -m flake8 .
+          python -m flake8 --extend-exclude tests/fixtures .
```
Same as the above comment. In fact I don't even think flake8 has this flag. This is probably actually obsolete.

Celleforst replied:
It does have this flag, at least it worked for me when testing locally. But I will remove it.

gn-kim-k replied:
But this is a remote CI job...

Strengthless commented on tests/fixtures/flake8_tests/trigger_F63/sample3.py:
_nit:_ for the future, it's probably better to name those `sample{int}.py` on a per-folder basis, e.g. we can have `sample1.py` and `sample2.py` for `trigger_E9`, then another set of `sample1.py` and `sample2.py` for `trigger_F63`. 

Otherwise, just imagine growing the test cases later on. 😄 Are you going to go through every folder just to check what ID we're currently at?

Strengthless commented on tests/test_actions.py:
```diff
@@ -1,9 +1,15 @@
     check_if_file_exists,
     check_if_folder_exists,
     remove_folder,
+    write_to_file,
```
_nit:_ we might want to remove this unused import.

Strengthless commented on tests/test_actions.py:
```diff
@@ -70,6 +76,32 @@ def test_setup_dependencies_failure(self):
             setup_dependencies(self.ephemeral_folder + "assignment-1")
 
+    # Tests for run_linter_check
+    def test_linting_without_errors(self):
+        check, _ = run_linter_check("tests/fixtures/flake8_tests/syntax_correct")
+        self.assertTrue(check)
```
Can we add one more TC here, with something like:
```py
user = "Cyrille"
print("Hello world")
```

This would create an unused variable, which isn't ideal (and should be caught by the linter in normal cases) - but we want to make sure that our function doesn't fail the CI in this case, since it should only serve as a syntax check.

Celleforst replied:
Interestingly it does not catch this when I run flake8 manually in the terminal... Might not be in the standard config (?). But I am including the test anyways.

Strengthless commented on tests/test_actions.py:
```diff
@@ -70,6 +76,32 @@ def test_setup_dependencies_failure(self):
+        self.assertTrue("F7" in logs)
+
+    # Did not manage to find examples that would trigger F63
+    # def test_linting_with_F63_errors(self):
+    #     check, logs = run_linter_check("tests/fixtures/flake8_tests/trigger_F63")
+    #     self.assertFalse(check)
+    #     self.assertTrue("F63" in logs)
```
For `F631`, we can reproduce using:
```py
assert (1 != 1, "1 is not equal to 1")
```

This asserts for a tuple, which probably returns true. The intended syntax is `assert 1 != 1, "1 is not equal to 1"`, which obviously is expected to fail as a test.

For `F632`, we can try this:
```py
if "cyrille" is "marcello":
    print("that's crazy!")
```

The intended syntax should be `if "cyrille" == "marcello"`.

Celleforst replied:
It worked, thanks!

Strengthless reviewed (changes requested):
Great coverage! Thanks for adding the tests 😄 just a few more things to fix

Strengthless commented on src/modules/actions.py:
```diff
@@ -78,11 +78,22 @@ def run_linter_check(target_folder: str) -> tuple[bool, str]:
+    if not check_if_folder_exists(target_folder):
+        raise ValueError(f"The provided path {target_folder} is not a valid directory.")
+
+    flake_command = "flake8 --select E9,F63,F82,F7 ."
```
One more note: I think we need to add `flake8` to our requirements.txt if we want this to run properly in the production environment.

Celleforst committed:
ref(#4): Add refactors desired in review

Removes unused write_to_file import in test_actions, renames samples in
fixtures folder, removes --extend-exclude flag form black.yml and
flake.yml github workflows

Celleforst committed:
fix(#4): Add F63 flake8 error test and syntax_correct case

As desired by reviewer, adds syntax correct case where there is an
unused variable which should not trigger our flake8 syntax check.
Besides that, adds samples that trigger F63 errors.

Celleforst committed:
fix(#4): Add flake8 to requirements

Adds flake8 to requirements.txt as desired by reviewer

Celleforst committed:
ref(#4): Improve log formatting

Improves formatting of log and adds more elegant check for success based
on returncode rather than checkin for substring in stdout.

Strengthless reviewed (approved):
Thanks!!

Celleforst committed:
Merge remote-tracking branch 'origin/main' into 4-add-linter-to-ci-server
